File: 2025/solutions/day07/task2.rs
use std::{
    fs::File,
    io::{self, BufRead},
    time::Instant,
};

const PATH: &str = "inputs/day07/input.txt";

fn task2() {
    let file = File::open(PATH).unwrap();
    let mut lines = io::BufReader::new(file).lines();

    let origin_line = lines.next().unwrap().unwrap();
    let idx = origin_line.find("S").unwrap();
    let mut frequencies = vec![0u64; origin_line.len()];
    frequencies[idx] = 1;

    for line in lines.map_while(Result::ok) {
        let mut next_frequencies = frequencies.clone();
        println!("{line}");
        for idx in 0..frequencies.len() {
            let frequency = frequencies[idx];
            if frequency == 0 {
                continue;
            }
            if &line[idx..(idx + 1)] == "^" {
                if idx > 0 {
                    next_frequencies[idx - 1] += frequency;
                }
                if idx < line.len() - 1 {
                    next_frequencies[idx + 1] += frequency;
                }
                next_frequencies[idx] -= frequency;
            }
        }
        println!("{:?}", next_frequencies);
        frequencies = next_frequencies;
    }

    let mut timelines = 0;
    for frequency in frequencies {
        timelines += frequency;
    }
    println!("{timelines}");
}

fn main() {
    let timer = Instant::now();
    task2();
    println!("Done in {:?}", timer.elapsed());
}

File: 2016/src/day15/task2/Task2.java
package day15.task2;

import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Arrays;

public class Task2 {
   Path path = Path.of("inputs", "day15", "input.txt");

   void task() {
      try (var lines = Files.lines(this.path)) {
         var discs = new ArrayList<>(lines.map((line) -> Arrays.stream(line
               .replace(" positions; at time=0, it is at position", "")
               .replace(".", "")
               .replaceAll(".*has ", "")
               .split(" "))
               .map((e) -> Integer.parseInt(e)).toList()).toList());
         discs.add(Arrays.asList(11, 0));

         var result = 0;
         timer: for (result = 0; true; result++) {
            for (var i = 0; i < discs.size(); i++) {
               var positions = discs.get(i).get(0);
               var start = discs.get(i).get(1);

               var currentPosition = (start + result + 1 + i) % positions;

               if (currentPosition != 0) {
                  continue timer;
               }
            }
            break;
         }

         System.out.println(result);
      } catch (Exception e) {
         e.printStackTrace();
         System.exit(1);
      }
   }

   public static void main(String[] args) {
      var start = Instant.now();

      new Task2().task();

      System.out.println("Done in " + Duration.between(start, Instant.now()).toString().substring(2).toLowerCase());
   }
}
